### ISSUE TITLE
GQ Signature flagging

### ISSUE THREAD

**Issue by EthanHeilman** [bastionzero/openpubkey#15]
Reading through https://github.com/bastionzero/openpubkey/pull/14 I had the following thoughts I wanted record about possible ways to flag that a signature if GQ Signature.

Consider the following two approaches:

The unsigned header flag approach
====

We flag that a OP RSA signature has been replaced by a GQ Signature by adding a unsigned header field: `"sig_type": "oidc_gq"` Lets call 

```json
{
	"payload": "payload",
	"signatures": [{
		"protected": {"alg":"ES256","rz":"5","upk":{"crv":"P-256","kty":"EC","x":"RI","y":"Hk"}},
		"header": {
			"sig_type": "cic"
		},
		"signature": "user signature"
	}, {
		"protected": {"alg":"RS256","kid":"6f7254101f56e41cf35c9926de84a2d552b4c6f1","typ":"JWT"},
		"header": {
			"sig_type": "oidc_gq"
		},
		"signature": "GQ signature"
	}]
}
```

The altered protected header approach
====

Alternatively we could alter the protected header changing the `alg` field to `GQ256` and then signing the altered protected header and payload with the GQ signature. A verifier would then reconstruct the message the OP signed with RSA by changing the algorithm field back. It would look like:

```json
{
	"payload": "payload",
	"signatures": [{
		"protected": {"alg":"ES256","rz":"5","upk":{"crv":"P-256","kty":"EC","x":"RI","y":"Hk"}},
		"header": {
			"sig_type": "cic"
		},
		"signature": "user signature"
	}, {
		"protected": {"alg":"GQ256","kid":"6","typ":"JWT"},
		"signature": "GQ signature"
	}]
}
```

Comparison
====

**The unsigned header flag approach**

**Pros:**
* Results in a simpler implementation of the the verifier and the signer.
* None of the values which are signed by OP are altered or reserialized.

**Cons:**
* Standard OIDC validators don't know about GQ signatures or the `sig_type` header so they will likely attempt to verify the GQ signature as if it was a very large RSA signature resulting in a hard to debug failure.
* Uses an unsigned header, which means we have to worry about what happens if someone alters this header. Ideally there is no content which is not covered by a signature. This is a minor con.

**The altered protected header approach**

**Pros:**
* Standard OIDC validators will see that the `alg` is set to algorithm they don't support and fail gracefully producing a meaningful error message: Signature algorithm `GQ256` is unsupported.
* If we want to standardize GQ signatures in JWS as `alg` is the standard way of flagging the signing algorithm

**Cons:**
* Likely a source of bugs and a pain point for implementors. Changing the protected header requires serializing and deserializing it. Different JSON libraries make minor changes to whitespace or the ordering of keys that result in signatures not verifying. These signature verification issues can creep in both when generating the GQ signature and when verifying it, as they both involve deserializing the protected header, making a change and then serializing it again.

We could bypass this con by storing the base64 encoding of the OPs signature in the protected header. That seems a little abusive toward the JWS standard.

**Comment by jonnystoten** [comment]
The altered protected header approach is very nice, but that JSON serialization con is a very real problem. My first GQ experiments had that exact bug :laughing: 

I was thinking we could just string replace - `s/"alg":"RS256"/"alg":"GQ256"/` - but TIL that the JSON inside a JWS can contain whitespace :unamused: 

> We could bypass this con by storing the base64 encoding of the OPs signature in the protected header. That seems a little abusive toward the JWS standard.

I'm not sure I follow this, could you rephrase?

**Comment by EthanHeilman** [comment]
> > We could bypass this con by storing the base64 encoding of the OPs signature in the protected header. That seems a little abusive toward the JWS standard.

> I'm not sure I follow this, could you rephrase?

OP RSA protected header (rsaph): `{"alg":"RSA256","kid":"6","typ":"JWT"}`

Take the OP's RSA protected header, base64 encode it and shove it into the GQ protected header, making the GQ protected header: `{"alg":"GQ256","typ":"JWT", "rsaph": base64({"alg":"RS256","kid":"6","typ":"JWT"})}`

**Comment by EthanHeilman** [comment]
I like the altered protected header approach. If someone wants to take it off the stack, let me know,. Otherwise I can put a PR together late next week.

**Comment by jonnystoten** [comment]
I've been looking at this and have some working code for the GQ256 alg approach. However, we've since started using the unprotected header approach for cosigner signatures as well.

I think it would be fine to simply move the claim signature type claim to the protected headers on cosigner signatures. At the moment the unprotected header is called `sig_type` but we might want something shorter. 

Any thoughts @EthanHeilman @lgmugnier ?

**Comment by EthanHeilman** [comment]
> I think it would be fine to simply move the claim signature type claim to the protected headers on cosigner signatures. At the moment the unprotected header is called sig_type but we might want something shorter.

One option would be to get rid of `sig_type` altogether and just inferring if a signature is OIDC, CIC or COS based on the protecter header parameters.

* OIDC signature - If no `upk` or `iss` claim in protected header.
* CIC signature - If `upk` in protected header.
* COS Signature - If `iss` in protected header.

If we did want to move the sig_type into the protected header, I would propose the using the `typ` (type) header introduced in [RFC-7515 JSON Web Signature (JWS)](https://www.rfc-editor.org/rfc/rfc7515.html#section-4.1.9) and used by [RFC-7519 JSON Web Token (JWT)](https://datatracker.ietf.org/doc/html/rfc7519#section-5.1) and defined in [the JOSE registry](https://www.iana.org/assignments/jose/jose.xhtml) in place of `sig_type`.

The beauty of this approach is that the OP's OIDC signature should already have `typ=JWT.` Technically the JWT RFC says, "Use of this Header Parameter is OPTIONAL", but I've yet to see any OIDC Provider that does not set `typ=JWT` in the protected header of their signature. I just checked and Google, Azure, and Github Actions all specify `typ=JWT.`

Even if an OP did not set `typ=JWT` in the protected header, we could infer a signature was the OIDC signature because either `typ=JWT` or no `typ` was set in the protected header.

* OIDC signature - If `typ=JWT` or no `typ` set
* CIC signature - if `typ=CIC`
* COS signature - if `typ=COS`

The only downside to this approach is that technically the `typ` header is suppose to describe the media type of the payload. We would be using it in a slightly off-label manner. The JWS RFC does really think through how typ should be used when there are multiple signatures and I think using typ in this fashion is reasonable and fits with the spirit if not the letter of the RFC.

RFC-7515 JSON Web Signature (JWS) Section 4.1.9](https://www.rfc-editor.org/rfc/rfc7515.html#section-4.1.9) 

> The "typ" (type) Header Parameter is used by JWS applications to declare the media type [[IANA.MediaTypes](https://www.rfc-editor.org/rfc/rfc7515.html#ref-IANA.MediaTypes)] of this complete JWS. This is intended for use by the application when more than one kind of object could be present in an application data structure that can contain a JWS; the application can use this value to disambiguate among the different kinds of objects that might be present.  It will typically not be used by applications when the kind of object is already known.  This parameter is ignored by JWS implementations; any processing of this parameter is performed by the JWS application. Use of this Header Parameter is OPTIONAL. [..]

[RFC-7519 JSON Web Token (JWT) Section 5.1](https://datatracker.ietf.org/doc/html/rfc7519#section-5.1)

> The "typ" (type) Header Parameter defined by [[JWS](https://datatracker.ietf.org/doc/html/rfc7519#ref-JWS)] and [[JWE](https://datatracker.ietf.org/doc/html/rfc7519#ref-JWE)] is used by JWT applications to declare the media type [[IANA.MediaTypes](https://datatracker.ietf.org/doc/html/rfc7519#ref-IANA.MediaTypes)] of this complete JWT.  This is intended for use by the JWT application when values that are not JWTs could also be present in an application data structure that can contain a JWT object; the application can use this value to disambiguate among the different kinds of objects that might be present.  It will typically not be used by applications when it is already known that the object is a JWT.  This parameter is ignored by JWT implementations; any processing of this parameter is performed by the JWT application.  If present, it is RECOMMENDED that its value be "JWT" to indicate that this object is a JWT.  While media type names are not case sensitive, it is RECOMMENDED that "JWT" always be spelled using uppercase characters for compatibility with legacy implementations.  Use of this Header Parameter is OPTIONAL.

**Comment by jonnystoten** [comment]
This sounds good, I've updated #66 to use this approach. Another option would have been `cty` (https://www.rfc-editor.org/rfc/rfc7515.html#section-4.1.10) but it is indeed nice that the OIDC tokens already have `typ` set to `JWT`.

**Comment by EthanHeilman** [comment]
>  Another option would have been cty

Oh, that is interesting. If I am reading the RFC correctly the cty only covers the "secured content" which means it doesn't refer to other signatures?

**Comment by jonnystoten** [comment]
I think it means it describes the payload field only, but (like with `typ`) it doesn't cover what it means to have different values in different signature headers.
